crates/wasmtime/src/engine.rs

```diff
@@ -217,6 +220,207 @@ impl Engine {
             .map(|a| f(a))
             .collect::<Result<Vec<B>, E>>()
     }
+
+    /// Returns the target triple which this engine is compiling code for
+    /// and/or running code for.
+    pub(crate) fn target(&self) -> target_lexicon::Triple {
+        // If a compiler is configured, use that target.
+        #[cfg(compiler)]
+        return self.compiler().triple().clone();
+
+        // ... otherwise it's the native target
+        #[cfg(not(compiler))]
+        return target_lexicon::Triple::host();
+    }
+
+    /// Verify that this engine's configuration is compatible with loading
+    /// modules onto the native host platform.
+    ///
+    /// This method is used as part of `Module::new` to ensure that this
+    /// engine can indeed load modules for the configured compiler (if any).
+    /// Note that if cranelift is disabled this trivially returns `Ok` because
+    /// loaded serialized modules are checked separately.
+    pub(crate) fn check_compatible_with_native_host(&self) -> Result<()> {
+        #[cfg(compiler)]
+        {
+            // Check if we've already performed this check
+            let flag = &self.inner.compatible_with_native_host;
```

Comment:
Can we use a `OnceCell` to implement the lazy computation here? Effectively we're computing either a boolean ("are we compatible") or a `Result`, depending on how you want to look at it; maybe a `OnceCell<Result>` to cache our return value?